### ISSUE TITLE
Jackson deserialization fails (Spring Boot 2.5)

### ISSUE THREAD

**Issue by sullrich84** [spring-projects/spring-data-rest#2023]
After migrating to Spring Boot 2.5, some of my controller tests failed with the following warning:

RepositoryRestMvcConfiguration$ResourceSupportHttpMessageConverter Failed to evaluate Jackson deserialization for type [[simple type, class org.springframework.data.rest.webmvc.PersistentEntityResource]]: java.lang.NullPointerException: Cannot invoke "org.springframework.data.util.TypeInformation.getActualType()" because "propertyType" is null

As far as I can tell, this only happens to tests where I place an URI reference in payloads, e.g.: `http://localhost/entity/foobar`. To me it looks like Spring Data Rest can't resolve the entity thats referenced by the URI. Any ideas what could be the issue?

### Test logs

```
10:58:23.048 [main] WARN RepositoryRestMvcConfiguration$ResourceSupportHttpMessageConverter Failed to evaluate Jackson deserialization for type [[simple type, class org.springframework.data.rest.webmvc.PersistentEntityResource]]: java.lang.NullPointerException: Cannot invoke "org.springframework.data.util.TypeInformation.getActualType()" because "propertyType" is null

MockHttpServletRequest:
      HTTP Method = POST
      Request URI = /follows
       Parameters = {}
          Headers = [Content-Type:"application/hal+json;charset=UTF-8", Content-Length:"66"]
             Body = {"followee":"http://localhost/athletes/paul","status":"REQUESTED"}
    Session Attrs = {}

Handler:
             Type = org.springframework.data.rest.webmvc.RepositoryEntityController
           Method = org.springframework.data.rest.webmvc.RepositoryEntityController#postCollectionResource(RootResourceInformation, PersistentEntityResource, PersistentEntityResourceAssembler, String)

Async:
    Async started = false
     Async result = null

Resolved Exception:
             Type = org.springframework.http.converter.HttpMessageNotReadableException

ModelAndView:
        View name = null
             View = null
            Model = null

FlashMap:
       Attributes = null

MockHttpServletResponse:
           Status = 400
    Error message = null
          Headers = [Vary:"Origin", "Access-Control-Request-Method", "Access-Control-Request-Headers", Content-Type:"application/vnd.wdb.problem+json;charset=UTF-8", X-Content-Type-Options:"nosniff", X-XSS-Protection:"1; mode=block", Cache-Control:"no-cache, no-store, max-age=0, must-revalidate", Pragma:"no-cache", Expires:"0", X-Frame-Options:"DENY"]
     Content type = application/vnd.wdb.problem+json;charset=UTF-8
             Body = {"status":400,"title":"Bad Request","detail":"Bad Request","instance":"http://localhost/follows","type":"https://developer.mozilla.org/docs/Web/HTTP/Status/400","invalid-params":[]}
    Forwarded URL = null
   Redirected URL = null
          Cookies = []

java.lang.AssertionError: Status expected:<201> but was:<400>
Expected :201
Actual   :400
```

Possible duplicate of #2015

**Comment by choryuidentify** [comment]
Same issue here.

AdoptOpenJDK 8u292
Spring Boot 2.5.0 (migrate from 2.1.1)
Spring Data Rest 3.5.1
MySQL 5.7

Failed to evaluate Jackson deserialization for type [[simple type, class org.springframework.data.rest.webmvc.PersistentEntityResource]]: java.lang.NullPointerException

It's maybe problem of processing List type.
My simple object without List are works good.

**Comment by sullrich84** [comment]
If you have HAL configured as the default MediaType, `jacksonHttpMessageConverter` is only registered for content types `application/schema+json`, `application/json-patch+json`, `application/merge-patch+json`, `application/x-spring-data-verbose+json` and `application/x-spring-data-compact+json`. 
 
 It looks like `application/hal+json` has been forgotten here.

**Comment by sullrich84** [comment]
Found out that there is a second converter call `halJacksonHttpMessageConverter` which should handle HAL media types. 

As far as I can tell, for `useHalAsDefaultJsonMediaType = true` scenarios, incoming `application/hal+json` or `application/json` payloads containing a link ref will be rejected with a 400.

For `useHalAsDefaultJsonMediaType = false` scenarios, incoming `application/json` payloads containing a link ref will be handled accordingly.

**Comment by mp911de** [comment]
Can you help us by providing a [minimal sample](https://stackoverflow.com/help/mcve) that reproduces the problem so we can make sure that we're on the right track? We currently see an increased number of reports that seem similar but might actually have different causes.

**Comment by dineshbhagat** [comment]
here it is: https://github.com/JanStureNielsen/so-68011276

**Comment by mjustin** [comment]
I'm running into the exact same error on upgrade in a slightly different when creating a MongoDB document (with Spring Data MongoDB) using POST, if the document has a nested object that is renaming a JSON property using the Jackson [`@JsonProperty`](http://fasterxml.github.io/jackson-annotations/javadoc/2.12/com/fasterxml/jackson/annotation/JsonProperty.html) annotation.  I do not get the error if the `@JsonProperty` is removed, or if the `@JsonProperty` rename is on the main document itself and not a nested object.

I assume this is the same issue, so I'll refrain from opening a separate issue unless I hear otherwise.

```java
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

@Document
public class MyDocument {
    @Id
    private String id;
    private NestedType value;

    public String getId() {return id;}
    public void setId(String id) {this.id = id;}
    public NestedType getValue() {return value;}
    public void setValue(NestedType value) {this.value = value;}
}
```

```java
import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.data.mongodb.core.mapping.Field;

public class NestedType {
    @Field("renamedValue")
    @JsonProperty("renamedValue")
    private String value;

    public String getValue() {return value;}
    public void setValue(String value) {this.value = value;}
}
```

**Comment by sullrich84** [comment]
This issue seems to be fixed when updating to Spring Boot 2.5.3.
